### DISCUSSION TITLE
on_failure_callback is not working if I put it inside dag_default_args?

### DISCUSSION THREAD

**Discussion by seaguest** [apache/airflow#30957] [General]
Hello, 

I found an inconsistent behavior when I tried to put on_failure_callback inside dag_default_args.
The following code works
```
#datetime
from datetime import timedelta, datetime
from airflow.operators.python import PythonVirtualenvOperator
import pendulum
from airflow import DAG
import logging

def on_failure_callback(context):
    logging.error("enter on_failure_callback-----------")

# python callable function
def print_hello():
    raise Exception("helloworld exception")

with DAG(
    dag_id="hello_world_dag",
    default_args={
        'on_failure_callback': on_failure_callback,
    },
    schedule_interval="00 02 * * *",
    start_date=pendulum.datetime(2022, 12, 1, tz="UTC"),
    catchup=False,
    max_active_runs=1,
) as dag:
    export = PythonVirtualenvOperator(
        python_callable=print_hello,
        task_id="hello",
        dag=dag,
    )
```

However this code doesn't work
```
#datetime
from datetime import timedelta, datetime
from airflow.operators.python import PythonVirtualenvOperator
import pendulum
from airflow import DAG
import logging

def on_failure_callback(context):
    logging.error("enter on_failure_callback-----------")

dag_default_args = {
    'retries': 1,
    'on_failure_callback': on_failure_callback,
}

# python callable function
def print_hello():
    raise Exception("helloworld exception")

with DAG(
    dag_id="hello_world_dag",
    default_args=dag_default_args,
    schedule_interval="00 02 * * *",
    start_date=pendulum.datetime(2022, 12, 1, tz="UTC"),
    catchup=False,
    max_active_runs=1,
) as dag:
    export = PythonVirtualenvOperator(
        python_callable=print_hello,
        task_id="hello",
        dag=dag,
    )
```

Why the second is not working?

**Comment by pankajastro** [comment]
Hi @seaguest what version of Airflow you are using? I tested this on main and look like it calls `on_failure_callback` when DAG fails. Yes, I see you have added `retries` in the 2nd DAG and it does not call `on_failure_callback` when still retying like when it fails 1st time in your case since you have retries=1 but does not this is expected and calls the on_failure_callback only if dag fail?

**Comment by seaguest** [comment]
I am using 2.5.3.
Indeed it works in second case if I remove retries.

**Reply by pankajastro** [reply]
It works even if you keep retries but `on_failure_callback` get called if the last retry fail
<img width="1490" alt="Screenshot 2023-04-29 at 8 41 55 PM" src="https://user-images.githubusercontent.com/98807258/235309971-7249e736-5afa-49b5-82c5-efc0031de4e2.png">

**Reply by pankajastro** [reply]
But I see a similar issue reported https://github.com/apache/airflow/discussions/30857 in 2.5.3 so maybe bug but working in main

**Reply by seaguest** [reply]
@pankajastro 
thanks, indeed the failure message only printed for the last retry, looks like on_failure_callback is triggered only for last retry.

**Reply by pankajastro** [reply]
FYI: we also have `on_retry_callback `  https://airflow.apache.org/docs/apache-airflow/2.6.0/administration-and-deployment/logging-monitoring/callbacks.html
